### PR TITLE
Fast-CDR external dependency Quality Declaration

### PR DESCRIPTION
**Pull request by Blast545**
This is a quality declaration for eProsima Fast-CDR external dependency for ROS2 packages as described in the proposed REP-2004.

This declaration represents some thoughts regarding the current state of this external dependency and how this affect Quality Level 1 for ROS2 packages, feedback is well received.

Signed-off-by: Jorge Perez <jjperez@ekumenlabs.com>

### CODE REVIEW

File: Quality_declaration_fastcdr.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+There is no coverage information summary/report and, apparently, no checks being made in terms of linters/static analysis, however, as the core parts used in ROS2 are actively being tested under the packages that use  `fast-cdr`.
+
+Considering the previously mentioned reasons, especially the fact that this software is being tested both directly and indirectly with [CI jobs](http://jenkins.eprosima.com:8080/), we consider this library to be robust and reliable, and hence we declare it to qualify as a level 1 external dependency.
+
+## Comparison with ROS packages quality standards
+
+### Version policy
+
+ 1. *Must have a version policy*
```
Comment (ahcorde) [inline Quality_declaration_fastcdr.md:20]: Use the Quality declaration template https://github.com/ros-infrastructure/rep/blob/4a3bbf6c296a725fda742ff8e695e2e1f6f59aa9/rep-2004.rst#quality-declaration-template

File: Quality_declaration_fastcdr.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+As stated in their public repository, *eProsima FastCDR is a C++ library that provides two serialization mechanisms. One is the standard CDR serialization mechanism, while the other is a faster implementation that modifies the standard*.
+
+## Summary
+The `fast-cdr` library is used as one of the dependencies of the `fastrtps`, it is important for the eProsima team to consider this software to be reliable. On its current state, meets most of the requirements to qualify as a level 1 package.
+
+As for the missing topics, eProsima does not state their versioning policies for the repository and does not state explicitly the API/ABI policies for the code in  `fast-cdr`. As the code deals with the standard RPTS implementation, and it is used by their software `fastrtps` it is not expected that the basic API functionalities will be changed. In any case, the code is actively being tested against their  [CI](http://jenkins.eprosima.com:8080/) (both directly with the included test cases and indirectly through the tests of `fastrtps`)  and indirectly in the [ROS2 CI](https://ci.ros2.org/).
```
Comment (ahcorde) [inline Quality_declaration_fastcdr.md:10]:
```suggestion
As for the missing topics, eProsima does not state their versioning policies for the repository and does not state explicitly the API/ABI policies for the code in `fast-cdr`. As the code deals with the standard RPTS implementation, and it is used by their software `fastrtps` it is not expected that the basic API functionalities will be changed. In any case, the code is actively being tested against their [CI](http://jenkins.eprosima.com:8080/) (both directly with the included test cases and indirectly through the tests of `fastrtps`) and indirectly in the [ROS2 CI](https://ci.ros2.org/).
```

File: Quality_declaration_fastcdr.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    
+There is no explicit policy related to API stability. This package should be pinned to a particular Fast-CDR version to be considered high quality.
+
+### ABI Stability Policy [1.v]
+    
+There is no explicit policy related to ABI stability. This package should be pinned to a particular Fast-CDR version to be considered high quality.
+
+### ABI and ABI Stability Within a Released ROS Distribution [1.vi]
+   
+There is no direct correlation between API and ABI stability of the library within ROS distributions. This package should be pinned to a particular Fast-CDR version to be considered high quality.
```
Comment (brawner) [inline Quality_declaration_fastcdr.md:40]: "Individual ROS distributions may need to choose specific versions to ensure API and ABI stability within a released ROS distribution"

File: Quality_declaration_fastcdr.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    
+Fast-CDR does not have a declared change control process. However, browsing through its public GitHub repository [history of commits](https://github.com/eProsima/Fast-CDR/commits/master) shows that most of the code passes through a Pull Request, except some minor cases related to styling issues or formatting.
+
+### Contributor Origin [2.ii]
+    
+Based on the public GitHub repository [history of commits](https://github.com/eProsima/Fast-CDR/commits/master) of `fast-cdr` it can be seen that there is no confirmation of contributor origin enforcement used in the repository.
+
+### Peer Review Policy [2.iii]
+    
+Not publicly stated, but browsing through their [list of closed pull requests](https://github.com/eProsima/Fast-CDR/pulls?q=is:pr%20is:closed) it can be seen that their workflow requires at least one peer review approving the changes and building against CI before merging code to the master branch in the repository.
```
Comment (brawner) [inline Quality_declaration_fastcdr.md:54]: "Fast-CDR does not have a publicly stated peer review policy"

File: Quality_declaration_fastcdr.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+### Continuous Integration [2.iv]
+    
+Not publicly stated, but browsing through their [list of closed pull requests](https://github.com/eProsima/Fast-CDR/pulls?q=is:pr%20is:closed) it can be seen that their workflow requires at least one peer review approving the changes and building against CI before merging code to the master branch in the repository.
```
Comment (brawner) [inline Quality_declaration_fastcdr.md:58]: "Fast-CDR does not have a publicly stated continuous integration process..."

File: Quality_declaration_fastcdr.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
+# **External Dependency Quality declaration** `fast-cdr` 
+
+This document is a declaration of software quality for the FastCDR external dependency, based on the guidelines in [REP-2004](https://github.com/ros-infrastructure/rep/blob/rep-2004/rep-2004.rst).
+
+The external dependency `fast-cdr` claims to be in the Quality Level 4 category.
```
Comment (brawner) [inline Quality_declaration_fastcdr.md:5]: Fast-CDR isn't exactly claiming this. I'd say something like, "This QD claims that the external dependency fast-cdr is ..."

File: Quality_declaration_fastcdr.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+As stated in their public repository, *eProsima FastCDR is a C++ library that provides two serialization mechanisms. One is the standard CDR serialization mechanism, while the other is a faster implementation that modifies the standard*.
+
+## Summary
+The `fast-cdr` library is used as one of the dependencies of the `fastrtps` library, it is important for the eProsima team to consider this software to be reliable. In its current state, meets most of the requirements to qualify as a level 1 package.
```
Comment (brawner) [inline Quality_declaration_fastcdr.md:10]:
```suggestion
The `fast-cdr` library is used as one of the dependencies of the `fastrtps` library, it is important for the eProsima team to consider this software to be reliable. In its current state, it meets most of the requirements to qualify as a level 1 package.
```

File: Quality_declaration_fastcdr.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+## Testing [4]
+
+### Feature Testing [4.i]
+    
+There are various [tests](https://github.com/eProsima/Fast-CDR/tree/master/test) in the `fast-cdr` repository, specifically the `SimpleTest.cpp` seems to be testing all the features of the library.
+
+### Public API Testing [4.ii]
+    
+There are several [tests](https://github.com/eProsima/Fast-CDR/tree/master/test) in the `fast-cdr` repository, specifically the `SimpleTest.cpp` seems to be testing all the functions/types/operations of the library.
```
Comment (brawner) [inline Quality_declaration_fastcdr.md:94]:
```suggestion
There are several [tests](https://github.com/eProsima/Fast-CDR/tree/master/test) in the `fast-cdr` repository, specifically `SimpleTest.cpp` seems to be testing all the functions/types/operations of the library.
```

File: Quality_declaration_fastcdr.md [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+Based on the [CI jobs test results](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/lastSuccessfulBuild/testReport/projectroot/test/), there seems there is no automatic static analysis being made on the code developed.
+
+## Dependencies [5]
+
+`fast-cdr` only depends on functions provided by the standard C++ definitions.
+
+## Platform Support [6]
+    
+There is no explicit support for any platform in their [repository page](https://github.com/eProsima/Fast-CDR). However, their CI jobs are triggered for both Linux and Windows. MacOS is not tested directly, but as this library is a dependence for the `Fast-RTPS` software, and that one is tested in the three platforms, it is safe to assume this library is supported in MacOS as well.
```
Comment (brawner) [inline Quality_declaration_fastcdr.md:116]:
```suggestion
There is no explicit support for any platform in their [repository page](https://github.com/eProsima/Fast-CDR). However, their CI jobs are triggered for both Linux and Windows. MacOS is not tested directly, but as this library is a dependency for the `Fast-RTPS` software, and that one is tested in the three platforms, it is safe to assume this library is supported in MacOS as well.
```

**Comment by ahcorde** [comment]
add the QD to the README.md

**Comment by EduPonz** [comment]
Due to Fast DDS reaching QL1 (eProsima/Fast-DDS#1610), I think this PR can be closed

**Pull request closed without merging**